File: Sources/JeffJS/Quantum/quantum_algorithms.py
#!/usr/bin/env python3
"""
quantum_algorithms.py — three quantum algorithms running on the
JeffJS stabilizer + Clifford-T simulator.

1. Deutsch-Jozsa: is f constant or balanced? (0 T gates, pure Clifford)
2. Bernstein-Vazirani: find the secret string s. (0 T gates, pure Clifford)
3. Toffoli (CCX) gate: universal computation via Clifford+T decomposition.
   (7 T gates — demonstrates the full CliffordTSimulator pipeline)

Each algorithm is implemented as a self-contained test function that
constructs the circuit, runs it, and verifies the correct result.

Run:
    python3 quantum_algorithms.py
"""

import numpy as np
import time
from stabilizer_sim import StabilizerState, CliffordTSimulator


# =====================================================================
# Algorithm 1: Deutsch-Jozsa
# =====================================================================

def deutsch_jozsa(n_input, oracle_type, oracle_bits=None):
    """Run the Deutsch-Jozsa algorithm.

    Given a function f: {0,1}^n → {0,1} that is PROMISED to be either:
      - constant (f(x) = 0 for all x, or f(x) = 1 for all x)
      - balanced (f(x) = 0 for exactly half of all x, f(x) = 1 for the other half)

    Determine which in ONE query.

    Circuit (n+1 qubits, the last is the output/ancilla):
      1. Set ancilla to |1⟩ (apply X)
      2. Apply H to all qubits
      3. Apply oracle U_f
      4. Apply H to input qubits
      5. Measure input qubits: all 0 → constant, any 1 → balanced

    Args:
        n_input: number of input qubits
        oracle_type: "constant_0", "constant_1", "balanced"
        oracle_bits: for balanced oracle, which input qubits CNOT to output
                     (e.g., [0, 2] means f depends on qubits 0 and 2)

    Returns:
        (result, measurements): "constant" or "balanced", and the measurement list
    """
    n_total = n_input + 1
    ancilla = n_input  # last qubit is the output/ancilla

    state = StabilizerState(n_total)

    # Step 1: Prepare ancilla in |1⟩
    state.x(ancilla)

    # Step 2: Hadamard all qubits
    for i in range(n_total):
        state.h(i)

    # Step 3: Oracle U_f
    if oracle_type == "constant_0":
        pass  # f(x) = 0 for all x → do nothing (identity)
    elif oracle_type == "constant_1":
        # f(x) = 1 for all x → flip the ancilla unconditionally
        state.x(ancilla)
    elif oracle_type == "balanced":
        # f(x) depends on specific input qubits → CNOT from each to ancilla
        if oracle_bits is None:
            oracle_bits = list(range(n_input))  # all bits → XOR function
        for q in oracle_bits:
            state.cnot(q, ancilla)

    # Step 4: Hadamard input qubits
    for i in range(n_input):
        state.h(i)

    # Step 5: Measure input qubits
    rng = np.random.default_rng(0xDECA)
    measurements = [state.measure(i, rng) for i in range(n_input)]

    # Interpretation
    if all(m == 0 for m in measurements):
        return "constant", measurements
    else:
        return "balanced", measurements


def test_deutsch_jozsa():
    """Test Deutsch-Jozsa on multiple oracle configurations."""
    print("=" * 60)
    print("Algorithm 1: DEUTSCH-JOZSA")
    print("Is f constant or balanced? Answer in 1 query.")
    print("Pure Clifford circuit — 0 T gates.")
    print("=" * 60)

    tests = [
        (4, "constant_0", None, "constant"),
        (4, "constant_1", None, "constant"),
        (4, "balanced", [0], "balanced"),
        (4, "balanced", [0, 1], "balanced"),
        (4, "balanced", [0, 1, 2, 3], "balanced"),
        (8, "constant_0", None, "constant"),
        (8, "balanced", [0, 2, 4, 6], "balanced"),
        (16, "balanced", [0, 3, 7, 15], "balanced"),
    ]

    all_pass = True
    print(f"\n  {'n':>4s} | {'oracle':>15s} | {'expected':>10s} | {'got':>10s} | {'pass':>5s}")
    print("  " + "-" * 55)

    for n, oracle_type, oracle_bits, expected in tests:
        result, meas = deutsch_jozsa(n, oracle_type, oracle_bits)
        ok = result == expected
        all_pass &= ok
        label = oracle_type
        if oracle_bits:
            label += f" {oracle_bits}"
        print(f"  {n:4d} | {label:>15s} | {expected:>10s} | {result:>10s} | {'YES' if ok else 'NO':>5s}")

    print(f"\n  {'ALL PASSED' if all_pass else 'SOME FAILED'}")
    return all_pass


# =====================================================================
# Algorithm 2: Bernstein-Vazirani
# =====================================================================

def bernstein_vazirani(secret_string):
    """Run the Bernstein-Vazirani algorithm.

    Given a function f(x) = s·x (mod 2) where s is a secret n-bit string,
    find s in ONE query.

    Circuit (n+1 qubits):
      1. Set ancilla to |1⟩
      2. Apply H to all qubits
      3. Oracle: CNOT from qubit i to ancilla for each bit of s that is 1
      4. Apply H to input qubits
      5. Measure input qubits → gives s directly

    Args:
        secret_string: string of '0' and '1', e.g., "10110"

    Returns:
        recovered_string: the measured string (should match secret_string)
    """
    n = len(secret_string)
    n_total = n + 1
    ancilla = n

    state = StabilizerState(n_total)

    # Step 1: Prepare ancilla
    state.x(ancilla)

    # Step 2: Hadamard all
    for i in range(n_total):
        state.h(i)

    # Step 3: Oracle — CNOT from qubit i to ancilla where s[i] = '1'
    for i, bit in enumerate(secret_string):
        if bit == '1':
            state.cnot(i, ancilla)

    # Step 4: Hadamard input qubits
    for i in range(n):
        state.h(i)

    # Step 5: Measure
    rng = np.random.default_rng(0xBEEF)
    measurements = [state.measure(i, rng) for i in range(n)]
    recovered = ''.join(str(m) for m in measurements)

    return recovered


def test_bernstein_vazirani():
    """Test Bernstein-Vazirani with various secret strings."""
    print("\n" + "=" * 60)
    print("Algorithm 2: BERNSTEIN-VAZIRANI")
    print("Find the secret string s in 1 query (classical needs n queries).")
    print("Pure Clifford circuit — 0 T gates.")
    print("=" * 60)

    secrets = [
        "101",
        "1111",
        "10110011",
        "0000000000000001",  # 16 bits, only last bit is 1
        "1010101010101010",  # 16 bits, alternating
        "11111111111111111111",  # 20 bits, all ones
    ]

    all_pass = True
    print(f"\n  {'n':>4s} | {'secret':>25s} | {'recovered':>25s} | {'pass':>5s}")
    print("  " + "-" * 70)

    for secret in secrets:
        recovered = bernstein_vazirani(secret)
        ok = recovered == secret
        all_pass &= ok
        n = len(secret)
        print(f"  {n:4d} | {secret:>25s} | {recovered:>25s} | {'YES' if ok else 'NO':>5s}")

    print(f"\n  {'ALL PASSED' if all_pass else 'SOME FAILED'}")
    return all_pass


# =====================================================================
# Algorithm 3: Toffoli gate (CCX) via Clifford+T decomposition
# =====================================================================

def toffoli_clifford_t(sim, q0, q1, q2):
    """Apply a Toffoli (controlled-controlled-NOT) gate using the standard
    Clifford+T decomposition. Uses 7 T/T† gates.

    Decomposition (Nielsen & Chuang, Fig 4.9):
      H(q2)
      CNOT(q1, q2); T†(q2)
      CNOT(q0, q2); T(q2)
      CNOT(q1, q2); T†(q2)
      CNOT(q0, q2); T(q1); T(q2)
      CNOT(q0, q1); H(q2)
      T(q0); T†(q1)
      CNOT(q0, q1)
    """
    # Standard decomposition: Barenco et al. 1995 / Nielsen & Chuang Fig 4.9
    # 6 CNOTs + 7 T/T† + 2 H = 15 gates total
    sim.h(q2)                # 1.  H on target
    sim.cnot(q1, q2)         # 2.  CNOT(control2, target)
    sim.t_dagger(q2)         # 3.  T† on target
    sim.cnot(q0, q2)         # 4.  CNOT(control1, target)
    sim.t(q2)                # 5.  T on target
    sim.cnot(q1, q2)         # 6.  CNOT(control2, target)
    sim.t_dagger(q2)         # 7.  T† on target
    sim.cnot(q0, q2)         # 8.  CNOT(control1, target)
    sim.t(q1)                # 9.  T on control2
    sim.t(q2)                # 10. T on target
    sim.h(q2)                # 11. H on target  ← was swapped with 12
    sim.cnot(q0, q1)         # 12. CNOT(control1, control2)
    sim.t(q0)                # 13. T on control1
    sim.t_dagger(q1)         # 14. T† on control2
    sim.cnot(q0, q1)         # 15. CNOT(control1, control2)


def test_toffoli():
    """Test the Toffoli gate on all 8 input combinations.
    The Toffoli gate flips qubit 2 if and only if qubits 0 AND 1 are both |1⟩.
    This is the quantum AND gate — it requires T gates (non-Clifford)
    and is the key to universal quantum computation.
    """
    print("\n" + "=" * 60)
    print("Algorithm 3: TOFFOLI GATE (CCX)")
    print("Controlled-controlled-NOT via Clifford+T decomposition.")
    print("7 T/T† gates → 2^7 = 128 stabilizer terms (universal computation).")
    print("=" * 60)

    rng = np.random.default_rng(0xCCCC)
    all_pass = True

    print(f"\n  {'input':>10s} | {'expected':>10s} | {'got':>10s} | {'terms':>6s} | {'time':>8s} | {'pass':>5s}")
    print("  " + "-" * 65)

    for q0_in in [0, 1]:
        for q1_in in [0, 1]:
            for q2_in in [0, 1]:
                # Expected: q2 flips iff q0=1 AND q1=1
                expected_q2 = q2_in ^ (q0_in & q1_in)
                expected = f"|{q0_in}{q1_in}{expected_q2}⟩"

                sim = CliffordTSimulator(3)

                # Prepare input state
                if q0_in:
                    sim.x(0)
                if q1_in:
                    sim.x(1)
                if q2_in:
                    sim.x(2)

                # Apply Toffoli
                t0 = time.time()
                toffoli_clifford_t(sim, 0, 1, 2)
                terms = sim.num_terms
                dt_gate = time.time() - t0

                # Measure
                t0 = time.time()
                m0 = sim.measure(0, rng)
                m1 = sim.measure(1, rng)
                m2 = sim.measure(2, rng)
                dt_meas = time.time() - t0

                got = f"|{m0}{m1}{m2}⟩"
                ok = (m0 == q0_in and m1 == q1_in and m2 == expected_q2)
                all_pass &= ok

                input_str = f"|{q0_in}{q1_in}{q2_in}⟩"
                print(f"  {input_str:>10s} | {expected:>10s} | {got:>10s} | {terms:>6d} | {dt_gate + dt_meas:7.3f}s | {'YES' if ok else 'NO':>5s}")

    print(f"\n  {'ALL 8 INPUTS CORRECT' if all_pass else 'SOME INPUTS FAILED'}")
    if all_pass:
        print("  >>> Toffoli gate works: universal quantum computation confirmed <<<")
    return all_pass


# =====================================================================
# Main
# =====================================================================

def main():
    print("*" * 60)
    print("JeffJS Quantum Algorithm Test Suite")
    print("Testing 3 quantum algorithms on the stabilizer + Clifford-T sim")
    print("*" * 60)

    t0 = time.time()

    ok1 = test_deutsch_jozsa()
    ok2 = test_bernstein_vazirani()
    ok3 = test_toffoli()

    total = time.time() - t0

    print("\n" + "=" * 60)
    print("SUMMARY")
    print("=" * 60)
    print(f"  Deutsch-Jozsa:         {'PASS' if ok1 else 'FAIL'}  (0 T gates, pure Clifford)")
    print(f"  Bernstein-Vazirani:    {'PASS' if ok2 else 'FAIL'}  (0 T gates, pure Clifford)")
    print(f"  Toffoli (CCX):         {'PASS' if ok3 else 'FAIL'}  (7 T gates, universal)")
    print(f"  Total time:            {total:.1f}s")
    print()

    if ok1 and ok2 and ok3:
        print("  >>> ALL ALGORITHMS PASS <<<")
        print("  The simulator correctly executes:")
        print("    - Clifford-only algorithms (Deutsch-Jozsa, Bernstein-Vazirani)")
        print("    - Universal circuits via Clifford+T decomposition (Toffoli)")
        print("  This is a working quantum computer simulator.")
    else:
        print("  !!! SOME ALGORITHMS FAILED !!!")


if __name__ == "__main__":
    main()

File: Sources/JeffJS/Quantum/QuantumAlgorithms.swift
// QuantumAlgorithms.swift
// Quantum algorithms implemented on the JeffJS QuantumCircuit API.
//
// Ported from quantum_algorithms.py and shor_iterative.py. Each algorithm
// is a static method on the QuantumAlgorithms enum. All circuits use
// QuantumCircuit for gate operations and deterministic SplitMix64 for
// measurement randomness.
//
// Algorithms:
//   1. Deutsch-Jozsa  — constant vs balanced, one query
//   2. Bernstein-Vazirani — recover secret string in one query
//   3. Toffoli (CCX) — Clifford+T decomposition (15 gates)
//   4. Shor's factoring — iterative phase estimation (n+1 qubits)
//   5. Quantum teleportation — transfer state via entanglement
//   6. Superdense coding — send 2 classical bits via 1 qubit

import Foundation

// MARK: - Result Types

/// Result of the Deutsch-Jozsa algorithm.
struct DeutschJozsaResult {
    let isConstant: Bool
    let measurements: [Int]
}

/// Oracle type for Deutsch-Jozsa.
enum OracleType {
    case constant0
    case constant1
    case balanced
}

/// Result of the Bernstein-Vazirani algorithm.
struct BernsteinVaziraniResult {
    let recoveredSecret: String
    let measurements: [Int]
}

/// Result of error correction.
struct ErrorCorrectionResult {
    let errorQubit: Int
    let syndrome: (Int, Int)
    let corrected: Bool
}

/// Result of Shor's factoring algorithm.
struct ShorResult {
    let factors: (Int, Int)?
    let N: Int
    let numQubits: Int
    let numTrials: Int
}

/// Result of quantum teleportation.
struct TeleportResult {
    let aliceMeasurements: (Int, Int)
    let bobOutcome: Int
    let success: Bool
}

/// Result of superdense coding.
struct SuperdenseResult {
    let sent: (Int, Int)
    let received: (Int, Int)
    let success: Bool
}

// MARK: - Helpers

/// Modular exponentiation: (base^exp) mod modulus, using repeated squaring.
/// Equivalent to Python's pow(base, exp, mod).
private func modPow(base: Int, exp: Int, mod: Int) -> Int {
    guard mod > 1 else { return 0 }
    var result = 1
    var b = base % mod
    var e = exp
    while e > 0 {
        if e & 1 == 1 {
            result = result * b % mod
        }
        e >>= 1
        b = b * b % mod
    }
    return result
}

/// Greatest common divisor (Euclidean algorithm).
private func gcd(_ a: Int, _ b: Int) -> Int {
    var a = abs(a)
    var b = abs(b)
    while b != 0 {
        let t = b
        b = a % b
        a = t
    }
    return a
}

/// Find the denominator of the closest fraction p/q to `value` with q <= maxDenom.
/// Uses the continued fraction expansion to converge on the best rational
/// approximation, which is what extracts the period in Shor's algorithm.
private func continuedFractionDenominator(of value: Double, maxDenom: Int) -> Int {
    guard value > 0 && value < 1 else { return 1 }

    var remainder = value
    var h0 = 0, h1 = 1
    var k0 = 1, k1 = 0

    for _ in 0 ..< 64 {
        let a = Int(floor(1.0 / remainder))
        let h2 = a * h1 + h0
        let k2 = a * k1 + k0

        if k2 > maxDenom { break }

        h0 = h1; h1 = h2
        k0 = k1; k1 = k2

        remainder = 1.0 / remainder - Double(a)
        if remainder < 1e-12 { break }
    }

    return k1 > 0 ? k1 : 1
}

/// Number of bits needed to represent N: ceil(log2(N+1)).
private func bitsNeeded(_ N: Int) -> Int {
    Int(ceil(log2(Double(N + 1))))
}

// MARK: - QuantumAlgorithms

/// Namespace for quantum algorithm implementations.
/// All methods are static; the enum has no cases (unconstructable).

enum QuantumAlgorithms {

    // MARK: 1. Deutsch-Jozsa

    /// Deutsch-Jozsa algorithm: determines whether an oracle function is
    /// constant or balanced using a single quantum query.
    ///
    /// - Parameters:
    ///   - numInputQubits: number of input qubits (n). Total circuit is n+1.
    ///   - oracle: the type of oracle to apply.
    ///   - oracleBits: for balanced oracles, which input qubits participate.
    ///     If nil, all input qubits are used.
    /// - Returns: result indicating constant/balanced and the measurement outcomes.
    static func deutschJozsa(
        numInputQubits: Int,
        oracle: OracleType,
        oracleBits: [Int]? = nil,
        seed: UInt64 = 0xDECA
    ) -> DeutschJozsaResult {
        let totalQubits = numInputQubits + 1
        let ancilla = numInputQubits
        let circuit = QuantumCircuit(numQubits: totalQubits, seed: seed)

        // Step 1: Prepare ancilla in |1>
        circuit.x(ancilla)

        // Step 2: Hadamard all qubits
        for q in 0 ..< totalQubits {
            circuit.h(q)
        }

        // Step 3: Apply oracle
        switch oracle {
        case .constant0:
            // f(x) = 0 for all x -> identity (do nothing)
            break
        case .constant1:
            // f(x) = 1 for all x -> flip the ancilla unconditionally
            circuit.x(ancilla)
        case .balanced:
            // f depends on specific input qubits -> CNOT from each to ancilla
            let bits = oracleBits ?? Array(0 ..< numInputQubits)
            for q in bits {
                circuit.cnot(control: q, target: ancilla)
            }
        }

        // Step 4: Hadamard input qubits only
        for q in 0 ..< numInputQubits {
            circuit.h(q)
        }

        // Step 5: Measure input qubits
        var measurements = [Int]()
        measurements.reserveCapacity(numInputQubits)
        for q in 0 ..< numInputQubits {
            measurements.append(circuit.measure(q))
        }

        // All zeros -> constant; any non-zero -> balanced
        let isConstant = measurements.allSatisfy { $0 == 0 }
        return DeutschJozsaResult(isConstant: isConstant, measurements: measurements)
    }

    // MARK: 2. Bernstein-Vazirani

    /// Bernstein-Vazirani algorithm: recovers a secret n-bit string s from
    /// a single query to an oracle computing f(x) = s . x (mod 2).
    ///
    /// - Parameter secretString: the secret string, e.g., "10110".
    /// - Returns: the recovered secret and raw measurements.
    static func bernsteinVazirani(
        secretString: String,
        seed: UInt64 = 0xBEEF
    ) -> BernsteinVaziraniResult {
        let n = secretString.count
        let totalQubits = n + 1
        let ancilla = n
        let circuit = QuantumCircuit(numQubits: totalQubits, seed: seed)

        // Step 1: Prepare ancilla in |1>
        circuit.x(ancilla)

        // Step 2: Hadamard all qubits
        for q in 0 ..< totalQubits {
            circuit.h(q)
        }

        // Step 3: Oracle -- CNOT from qubit i to ancilla where s[i] == '1'
        let chars = Array(secretString)
        for (i, bit) in chars.enumerated() {
            if bit == "1" {
                circuit.cnot(control: i, target: ancilla)
            }
        }

        // Step 4: Hadamard input qubits
        for q in 0 ..< n {
            circuit.h(q)
        }

        // Step 5: Measure input qubits
        var measurements = [Int]()
        measurements.reserveCapacity(n)
        for q in 0 ..< n {
            measurements.append(circuit.measure(q))
        }

        let recovered = measurements.map { String($0) }.joined()
        return BernsteinVaziraniResult(recoveredSecret: recovered, measurements: measurements)
    }

    // MARK: 3. Toffoli (CCX)

    /// Apply a Toffoli (controlled-controlled-NOT) gate using the standard
    /// Clifford+T decomposition (Barenco et al. 1995 / Nielsen & Chuang Fig 4.9).
    ///
    /// 6 CNOTs + 7 T/T-dagger + 2 H = 15 gates total.
    /// Flips q2 if and only if both q0 and q1 are |1>.
    static func toffoli(circuit: QuantumCircuit, q0: Int, q1: Int, q2: Int) {
        circuit.h(q2)                              //  1. H on target
        circuit.cnot(control: q1, target: q2)      //  2. CNOT(control2, target)
        circuit.tDagger(q2)                        //  3. T-dagger on target
        circuit.cnot(control: q0, target: q2)      //  4. CNOT(control1, target)
        circuit.t(q2)                              //  5. T on target
        circuit.cnot(control: q1, target: q2)      //  6. CNOT(control2, target)
        circuit.tDagger(q2)                        //  7. T-dagger on target
        circuit.cnot(control: q0, target: q2)      //  8. CNOT(control1, target)
        circuit.t(q1)                              //  9. T on control2
        circuit.t(q2)                              // 10. T on target
        circuit.h(q2)                              // 11. H on target
        circuit.cnot(control: q0, target: q1)      // 12. CNOT(control1, control2)
        circuit.t(q0)                              // 13. T on control1
        circuit.tDagger(q1)                        // 14. T-dagger on control2
        circuit.cnot(control: q0, target: q1)      // 15. CNOT(control1, control2)
    }

    // MARK: 4. Shor's Algorithm (Iterative Phase Estimation)

    /// Shor's factoring algorithm using iterative (semi-classical) phase estimation.
    ///
    /// Uses only n+1 qubits (1 counting + n work) where n = ceil(log2(N+1)).
    /// The counting qubit is measured and recycled 2n times with classical
    /// phase corrections replacing the full QFT.
    ///
    /// - Parameters:
    ///   - N: the number to factor (must be odd and composite).
    ///   - maxTrials: maximum number of random bases to try.
    ///   - seed: PRNG seed for deterministic results.
    /// - Returns: ShorResult with factors (if found), qubit count, and trial count.
    static func shorFactor(_ N: Int, maxTrials: Int = 30, seed: UInt64 = 0xDEAD) -> ShorResult {
        // Handle trivial even case
        if N % 2 == 0 {
            return ShorResult(factors: (2, N / 2), N: N, numQubits: 2, numTrials: 0)
        }

        let nBits = bitsNeeded(N)
        let m = 2 * nBits                // phase estimation rounds
        let nWork = nBits
        let nTotal = nWork + 1           // 1 counting + n work qubits
        let countingQubit = 0
        let workOffset = 1

        // Separate RNG for choosing random bases
        var baseRng = SplitMix64(seed: 42 + UInt64(N))

        for trial in 0 ..< maxTrials {
            // Pick a random base a in [2, N-1]
            let range = UInt64(N - 2)
            let a = Int(baseRng.next() % range) + 2
            let g = gcd(a, N)
            if g > 1 {
                // Lucky: gcd(a, N) is already a factor
                return ShorResult(factors: (g, N / g), N: N, numQubits: nTotal, numTrials: trial + 1)
            }

            // Build the circuit: 1 counting qubit + nWork work qubits
            let circuit = QuantumCircuit(numQubits: nTotal, seed: seed &+ UInt64(trial))

            // Initialize work register to |1> (qubit at workOffset = LSB of work reg)
            circuit.x(workOffset)

            var measuredBits = [Int]()
            measuredBits.reserveCapacity(m)

            for roundIdx in 0 ..< m {
                let j = m - 1 - roundIdx    // phase estimation index (MSB first)

                // Reset counting qubit to |0> by flipping if last measurement was 1
                if let lastBit = measuredBits.last, lastBit == 1 {
                    circuit.x(countingQubit)
                }

                // H on counting qubit -> |+>
                circuit.h(countingQubit)

                // Controlled-U^(2^j): modular multiplication by a^(2^j) mod N
                let power = modPow(base: a, exp: 1 << j, mod: N)
                if power != 1 {
                    circuit.controlledModMult(
                        control: countingQubit,
                        workOffset: workOffset,
                        nWork: nWork,
                        a: UInt32(power),
                        N: UInt32(N)
                    )
                }

                // Phase correction from previously measured bits (semi-classical QFT)
                var correction: Float = 0.0
                for (prevIdx, bk) in measuredBits.enumerated() {
                    let distance = roundIdx - prevIdx
                    if bk == 1 {
                        correction -= 2.0 * .pi / Float(1 << (distance + 1))
                    }
                }
                if abs(correction) > 1e-10 {
                    circuit.phase(countingQubit, angle: correction)
                }

                // H on counting qubit
                circuit.h(countingQubit)

                // Measure counting qubit
                let outcome = circuit.measure(countingQubit)
                measuredBits.append(outcome)
            }

            // Construct phase estimate from measured bits (MSB first)
            var phaseInt = 0
            for (idx, b) in measuredBits.enumerated() {
                phaseInt |= (b << (m - 1 - idx))
            }
            let phaseValue = Double(phaseInt) / Double(1 << m)

            // Use continued fractions to find the period
            guard phaseValue > 0 else { continue }
            let r = continuedFractionDenominator(of: phaseValue, maxDenom: N)

            // Free the circuit's memory before trying to extract factors
            circuit.deallocate()

            // Try to extract factors from the period
            if r > 0 && r % 2 == 0 {
                let halfPow = modPow(base: a, exp: r / 2, mod: N)
                let candidates = [gcd(halfPow - 1, N), gcd(halfPow + 1, N)]
                for g in candidates {
                    if g > 1 && g < N {
                        return ShorResult(
                            factors: (g, N / g),
                            N: N,
                            numQubits: nTotal,
                            numTrials: trial + 1
                        )
                    }
                }
            }
        }

        // No factors found within maxTrials
        return ShorResult(factors: nil, N: N, numQubits: nTotal, numTrials: maxTrials)
    }

    // MARK: 5. Quantum Teleportation

    /// Quantum teleportation: transfers a quantum state from one qubit to
    /// another using entanglement and classical communication.
    ///
    /// Protocol:
    ///   - Qubit 0 (data): prepared in |+> (the state to teleport)
    ///   - Qubit 1 (alice): half of a Bell pair
    ///   - Qubit 2 (bob): other half of the Bell pair
    ///
    /// After teleportation, Bob's qubit should be in the |+> state.
    /// Verification: measure Bob in the X basis (H then measure) -- should get 0.
    static func teleport(seed: UInt64 = 0x7E1E) -> TeleportResult {
        let circuit = QuantumCircuit(numQubits: 3, seed: seed)
        let data = 0
        let alice = 1
        let bob = 2

        // Prepare data qubit in |+> state
        circuit.h(data)

        // Create Bell pair between alice and bob
        circuit.h(alice)
        circuit.cnot(control: alice, target: bob)

        // Alice's operations: entangle data with her half of the Bell pair
        circuit.cnot(control: data, target: alice)
        circuit.h(data)

        // Alice measures both her qubits
        let m0 = circuit.measure(data)
        let m1 = circuit.measure(alice)

        // Bob applies corrections based on Alice's measurements
        if m1 == 1 {
            circuit.x(bob)      // correct for alice's bit flip
        }
        if m0 == 1 {
            circuit.z(bob)      // correct for data's phase flip
        }

        // Verify Bob has |+> by measuring in X basis (apply H then measure)
        circuit.h(bob)
        let bobResult = circuit.measure(bob)

        // Success if Bob measures 0 (meaning he was in |+> before the H)
        let success = (bobResult == 0)
        return TeleportResult(
            aliceMeasurements: (m0, m1),
            bobOutcome: bobResult,
            success: success
        )
    }

    // MARK: 6. Superdense Coding

    /// Superdense coding: transmit 2 classical bits by sending 1 qubit,
    /// using a pre-shared Bell pair.
    ///
    /// Protocol:
    ///   1. Create Bell pair (qubit 0 to Alice, qubit 1 to Bob).
    ///   2. Alice encodes 2 bits by applying gates to her qubit:
    ///      - (0,0): identity
    ///      - (0,1): X
    ///      - (1,0): Z
    ///      - (1,1): X then Z (equivalently, iY)
    ///   3. Alice sends her qubit to Bob.
    ///   4. Bob applies CNOT(alice, bob) then H(alice), then measures both.
    ///   5. Measurement results = the two sent bits.
    static func superdenseCoding(
        bit0: Int,
        bit1: Int,
        seed: UInt64 = 0x5D5D
    ) -> SuperdenseResult {
        let circuit = QuantumCircuit(numQubits: 2, seed: seed)
        let aliceQubit = 0
        let bobQubit = 1

        // Create Bell pair |00> + |11> (unnormalized)
        circuit.h(aliceQubit)
        circuit.cnot(control: aliceQubit, target: bobQubit)

        // Alice encodes 2 classical bits
        if bit1 == 1 {
            circuit.x(aliceQubit)
        }
        if bit0 == 1 {
            circuit.z(aliceQubit)
        }

        // Bob decodes: CNOT then H
        circuit.cnot(control: aliceQubit, target: bobQubit)
        circuit.h(aliceQubit)

        // Measure both qubits
        let r0 = circuit.measure(aliceQubit)
        let r1 = circuit.measure(bobQubit)

        let success = (r0 == bit0 && r1 == bit1)
        return SuperdenseResult(
            sent: (bit0, bit1),
            received: (r0, r1),
            success: success
        )
    }

    // MARK: - Error Correction (3-qubit bit-flip code)

    /// Encode 1 logical qubit into 3 physical qubits (bit-flip code),
    /// introduce an error on one qubit, detect and correct it.
    /// Returns whether the correction succeeded.
    static func errorCorrection(errorQubit: Int = 1, seed: UInt64 = 0xEC01) -> ErrorCorrectionResult {
        // 5 qubits: 0 = data, 1-2 = encoded copies, 3-4 = syndrome ancillas
        let circuit = QuantumCircuit(numQubits: 5, seed: seed)

        // Prepare data qubit in |+⟩ for a non-trivial test
        circuit.h(0)

        // Encode: CNOT from data to copies (bit-flip repetition code)
        // |ψ⟩ → |ψψψ⟩ (logical qubit spread across qubits 0, 1, 2)
        circuit.cnot(control: 0, target: 1)
        circuit.cnot(control: 0, target: 2)

        // Introduce a bit-flip error on the specified qubit
        let errorTarget = min(max(errorQubit, 0), 2)
        circuit.x(errorTarget)

        // Syndrome extraction using ancilla qubits 3, 4
        // Ancilla 3 = parity of qubits 0 and 1
        circuit.cnot(control: 0, target: 3)
        circuit.cnot(control: 1, target: 3)
        // Ancilla 4 = parity of qubits 1 and 2
        circuit.cnot(control: 1, target: 4)
        circuit.cnot(control: 2, target: 4)

        // Measure syndrome ancillas
        let s0 = circuit.measure(3)
        let s1 = circuit.measure(4)

        // Decode syndrome and apply correction
        // s0=0, s1=0 → no error
        // s0=1, s1=0 → error on qubit 0
        // s0=1, s1=1 → error on qubit 1
        // s0=0, s1=1 → error on qubit 2
        if s0 == 1 && s1 == 0 {
            circuit.x(0)  // correct qubit 0
        } else if s0 == 1 && s1 == 1 {
            circuit.x(1)  // correct qubit 1
        } else if s0 == 0 && s1 == 1 {
            circuit.x(2)  // correct qubit 2
        }

        // Decode: reverse the encoding
        circuit.cnot(control: 0, target: 2)
        circuit.cnot(control: 0, target: 1)

        // Verify: measure data qubit in X basis (should get |+⟩ = 0 after H)
        circuit.h(0)
        let dataResult = circuit.measure(0)

        // If correction worked, data qubit should be |+⟩ → measure 0 after H
        let success = (dataResult == 0)

        return ErrorCorrectionResult(
            errorQubit: errorTarget,
            syndrome: (s0, s1),
            corrected: success
        )
    }
}
